File: esm-front/src/app/services/grade.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Grade {
  id?: number;
  assessmentId: number;
  studentName: string;
  studentEmail: string;
  score: number;
  maxScore: number;
  comments?: string;
  gradedAt?: string;
  percentage?: number;
  mention?: string;
}

export interface GradeStats {
  total: number;
  average: number;
  max: number;
  min: number;
  passing: number;
  failing: number;
  passRate: number;
}

export interface LeaderboardEntry {
  rank: number;
  studentName: string;
  studentEmail: string;
  averagePercentage?: number;
  averageScore?: number;
  totalAssessments?: number;
  score?: number;
  maxScore?: number;
  percentage?: number;
  mention: string;
  comments?: string;
}

@Injectable({ providedIn: 'root' })
export class GradeService {

  private api = 'http://localhost:8080/api/grades';

  constructor(private http: HttpClient) { }

  getAll(): Observable<Grade[]> {
    return this.http.get<Grade[]>(this.api);
  }

  getById(id: number): Observable<Grade> {
    return this.http.get<Grade>(`${this.api}/${id}`);
  }

  getByAssessment(assessmentId: number): Observable<Grade[]> {
    return this.http.get<Grade[]>(`${this.api}/assessment/${assessmentId}`);
  }

  getByStudent(email: string): Observable<Grade[]> {
    return this.http.get<Grade[]>(`${this.api}/student/${email}`);
  }

  getStatsByAssessment(assessmentId: number): Observable<GradeStats> {
    return this.http.get<GradeStats>(`${this.api}/assessment/${assessmentId}/stats`);
  }

  getGlobalLeaderboard(): Observable<LeaderboardEntry[]> {
    return this.http.get<LeaderboardEntry[]>(`${this.api}/leaderboard`);
  }

  getLeaderboardByAssessment(assessmentId: number): Observable<LeaderboardEntry[]> {
    return this.http.get<LeaderboardEntry[]>(`${this.api}/leaderboard/assessment/${assessmentId}`);
  }

  create(grade: Grade): Observable<Grade> {
    return this.http.post<Grade>(this.api, grade);
  }

  update(id: number, grade: Grade): Observable<Grade> {
    return this.http.put<Grade>(`${this.api}/${id}`, grade);
  }

  delete(id: number): Observable<any> {
    return this.http.delete(`${this.api}/${id}`, { observe: 'response' });
  }
}

File: backend/assessment-service/src/main/java/com/example/assessmentservice/service/GradeService.java
package com.example.assessmentservice.service;

import com.example.assessmentservice.entity.Grade;
import com.example.assessmentservice.repository.GradeRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class GradeService {

    private final GradeRepository gradeRepository;

    // ── CRUD ───────────────────────────────────────────────────────────────────
    public Grade create(Grade grade) {
        return gradeRepository.save(grade);
    }

    public List<Grade> getAll() {
        return gradeRepository.findAll();
    }

    public Grade getById(Long id) {
        return gradeRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Grade not found: " + id));
    }

    public List<Grade> getByAssessment(Long assessmentId) {
        return gradeRepository.findByAssessmentId(assessmentId);
    }

    public List<Grade> getByStudent(String studentEmail) {
        return gradeRepository.findByStudentEmail(studentEmail);
    }

    public Grade update(Long id, Grade updated) {
        Grade existing = getById(id);
        existing.setStudentName(updated.getStudentName());
        existing.setStudentEmail(updated.getStudentEmail());
        existing.setScore(updated.getScore());
        existing.setMaxScore(updated.getMaxScore());
        existing.setComments(updated.getComments());
        return gradeRepository.save(existing);
    }

    public void delete(Long id) {
        gradeRepository.deleteById(id);
    }

    // ── Statistics ─────────────────────────────────────────────────────────────
    public Map<String, Object> getStatsByAssessment(Long assessmentId) {
        List<Grade> grades = gradeRepository.findByAssessmentId(assessmentId);
        long total   = grades.size();
        long passing = gradeRepository.countPassing(assessmentId);

        Map<String, Object> stats = new HashMap<>();
        stats.put("total",    total);
        stats.put("average",  gradeRepository.getAverageByAssessment(assessmentId));
        stats.put("max",      gradeRepository.getMaxByAssessment(assessmentId));
        stats.put("min",      gradeRepository.getMinByAssessment(assessmentId));
        stats.put("passing",  passing);
        stats.put("failing",  total - passing);
        stats.put("passRate", total > 0 ? Math.round((passing * 100.0) / total) : 0);
        return stats;
    }

    // ── Leaderboard ────────────────────────────────────────────────────────────

    /**
     * Global leaderboard — average percentage across ALL assessments per student
     */
    public List<Map<String, Object>> getGlobalLeaderboard() {
        List<Grade> all = gradeRepository.findAll();

        // Group by student email
        Map<String, List<Grade>> byStudent = all.stream()
                .collect(Collectors.groupingBy(Grade::getStudentEmail));

        List<Map<String, Object>> leaderboard = new ArrayList<>();

        byStudent.forEach((email, grades) -> {
            double totalPct = grades.stream()
                    .filter(g -> g.getMaxScore() != null && g.getMaxScore() > 0)
                    .mapToDouble(g -> (g.getScore() / g.getMaxScore()) * 100)
                    .average()
                    .orElse(0.0);

            double avgScore = grades.stream()
                    .mapToDouble(Grade::getScore)
                    .average()
                    .orElse(0.0);

            Map<String, Object> entry = new LinkedHashMap<>();
            entry.put("studentName",       grades.get(0).getStudentName());
            entry.put("studentEmail",      email);
            entry.put("averagePercentage", Math.round(totalPct * 10.0) / 10.0);
            entry.put("averageScore",      Math.round(avgScore * 10.0) / 10.0);
            entry.put("totalAssessments",  grades.size());
            entry.put("mention",           getMention(totalPct));
            leaderboard.add(entry);
        });

        // Sort by averagePercentage descending
        leaderboard.sort((a, b) ->
                Double.compare((Double) b.get("averagePercentage"), (Double) a.get("averagePercentage")));

        // Add rank
        for (int i = 0; i < leaderboard.size(); i++) {
            leaderboard.get(i).put("rank", i + 1);
        }

        return leaderboard;
    }

    /**
     * Leaderboard for a specific assessment
     */
    public List<Map<String, Object>> getLeaderboardByAssessment(Long assessmentId) {
        List<Grade> grades = gradeRepository.findByAssessmentId(assessmentId);

        List<Map<String, Object>> leaderboard = grades.stream()
                .sorted(Comparator.comparingDouble(Grade::getScore).reversed())
                .map(g -> {
                    double pct = g.getMaxScore() != null && g.getMaxScore() > 0
                            ? (g.getScore() / g.getMaxScore()) * 100 : 0;
                    Map<String, Object> entry = new LinkedHashMap<>();
                    entry.put("studentName",  g.getStudentName());
                    entry.put("studentEmail", g.getStudentEmail());
                    entry.put("score",        g.getScore());
                    entry.put("maxScore",     g.getMaxScore());
                    entry.put("percentage",   Math.round(pct * 10.0) / 10.0);
                    entry.put("mention",      getMention(pct));
                    entry.put("comments",     g.getComments());
                    return entry;
                })
                .collect(Collectors.toList());

        // Add rank
        for (int i = 0; i < leaderboard.size(); i++) {
            leaderboard.get(i).put("rank", i + 1);
        }

        return leaderboard;
    }

    private String getMention(double percentage) {
        if (percentage >= 90) return "EXCELLENT";
        if (percentage >= 75) return "GOOD";
        if (percentage >= 60) return "AVERAGE";
        return "FAIL";
    }
}